art/defences/preprocessor/spatial_smoothing_pytorch.py

```diff
@@ -0,0 +1,213 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the local spatial smoothing defence in `SpatialSmoothing` in PyTorch.
+
+| Paper link: https://arxiv.org/abs/1704.01155
+
+
+| Please keep in mind the limitations of defences. For more information on the limitations of this defence,
+    see https://arxiv.org/abs/1803.09868 . For details on how to evaluate classifier security in general, see
+    https://arxiv.org/abs/1902.06705
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+from kornia.filters.kernels import get_binary_kernel2d
+import torch.nn.functional as F
+import torch.nn as nn
+
+import logging
+from typing import Optional, Tuple
+
+import numpy as np
+import torch
+from kornia.filters import MedianBlur
```

Comment:
Please move PyTorch and korna related imports inside the method using them. This is necessary to allow using ART without PyTorch.

Reply:
I can do that. But I will have to import torch repeatedly within a file. Please let me know if you have a simpler workaround.

Reply:
I agree it is not the best solution, but it's the current approach of ART to support multiple frameworks without requiring to install all of them. We have been thinking about changing/improving it, but we haven't implemented it yet.

Reply:
Would `try-except` work here?

```python
try:
    import torch
except ImportError:
    print("torch not found.")
```

Reply:
I think this would probably work, but it would be a new pattern for ART. So far ART has always included the imports inside of the methods. Basically we need to be able to import `art` in an environment where a framework, in this case `torch`, is not installed.

I'm not sure if we should do the print, because if we would follow this new pattern in all files of ART it could lead to a lot of printing when importing `art` in an environment that has only a minimal dependency installation.

Reply:
I have made modifications following the previous practices in ART.

Shall we replace `print()` with `logger.debug()` or something like that? I think the `try...except...` style should help to simplify the code.